Automattic/newspack-newsletters
# feat: send list classes

### All Submissions:






### Changes proposed in this Pull Request:



Implements a new `Send_List` class for abstacting ESP-specific list and sublist data into a single schema. Also implements a `Send_Lists` class for shared util methods and REST API routes (the latter currently won't work because this PR doesn't include the `get_send_lists()` methods on each provider class).

To support the new REST API routes, moves the `newsletters/v1` namespace and the API permissions callback to the main `Newspack_Newsletters` class, so these can be shared outside of the `Newspack_Newsletters_Subscription` class.

### How to test the changes in this Pull Request:

1. Confirm that all tests pass (and review the tests for expected `Send_List` behavior)
2. In `wp_shell`, try the new constructor:

```bash
> $send_list = new Newspack\Newsletters\Send_List( [ 'id' => '123', 'name' => 'My Send List', 'type' => 'list', 'entity_type' => 'something', 'count' => 100, 'provider' => 'mailchimp' ] )
= Newspack\Newsletters\Send_List {#14997
    +"provider": "mailchimp",
    +"type": "list",
    +"entity_type": "something",
    +"id": "123",
    +"name": "My Send List",
    +"count": 100,
    +"value": "123",
    +"label": "[SOMETHING] My Send List (100 contacts)",
  }
```

3. Try out the new methods:

```bash
> $send_list->get_id()
= "123"

> $send_list->get_label()
= "[SOMETHING] My Send List (100 contacts)"

> $send_list->set_name( 'My Send List (renamed)' )
= "My Send List (renamed)"

> $send_list->get_name()
= "My Send List (renamed)"

> Newspack\Newsletters\Send_Lists::matches_search( 'renamed', [ $send_list->get( 'name' ) ] )
= true

> Newspack\Newsletters\Send_Lists::matches_id( [ 123 ], $send_list->get( 'id' ) )
= true

> $send_list->to_array()
= [
    "provider" => "mailchimp",
    "type" => "list",
    "entity_type" => "something",
    "id" => "123",
    "name" => "My Send List",
    "count" => 100,
    "label" => "[SOMETHING] My Send List (100 contacts)",
    "value" => "123",
  ]
```

### Other information:





---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207992056607959
